frame/dex-router/src/lib.rs

```diff
@@ -130,31 +126,47 @@ pub mod pallet {
 	impl<T: Config> Pallet<T> {}
 
 	impl<T: Config> Pallet<T> {
+		fn check_route(
+			asset_pair: CurrencyPair<T::AssetId>,
+			route: &BoundedVec<DexRouteNode<T::PoolId>, T::MaxHopsInRoute>,
+		) -> Result<(), DispatchError> {
+			ensure!(route.len() > 1, Error::<T>::MoreThanOneNodesExpectedInRoute);
+			let get_currency_pair =
+				|r: &DexRouteNode<T::PoolId>| -> Result<CurrencyPair<T::AssetId>, DispatchError> {
+					match r {
+						DexRouteNode::Pablo(pool_id) => T::Pablo::currency_pair(*pool_id),
+					}
+				};
+			// check first node's quote asset is same as quote asset of asset_pair.
+			let currency_pair = get_currency_pair(
+				route.first().ok_or(Error::<T>::MoreThanOneNodesExpectedInRoute)?,
+			)?;
+			ensure!(currency_pair.quote == asset_pair.quote, Error::<T>::UnexpectedNodeFound);
+			// check last node's base asset is same as base asset of asset_pair.
+			let currency_pair = get_currency_pair(
+				route.last().ok_or(Error::<T>::MoreThanOneNodesExpectedInRoute)?,
+			)?;
+			ensure!(currency_pair.base == asset_pair.base, Error::<T>::UnexpectedNodeFound);
+			// check base asset of last node is quote asset of current node, i.e route must have
+			// transitivity
+			let mut last_base = asset_pair.quote;
+			for r in route.as_slice() {
+				let currency_pair = get_currency_pair(r)?;
+				ensure!(last_base == currency_pair.quote, Error::<T>::UnexpectedNodeFound);
+				last_base = currency_pair.base;
+			}
+			Ok(())
```

Comment:
You could make this a lot easier with a try_fold, starting with an error and then shifting to OK if transitivity is respected